microcosm_eventsource/controllers.py

```diff
@@ -0,0 +1,134 @@
+"""
+Event controllers.
+
+"""
+from inflection import camelize
+
+from microcosm_flask.conventions.crud_adapter import CRUDStoreAdapter
+from microcosm_flask.conventions.encoding import with_context
+from microcosm_flask.naming import name_for
+from microcosm_flask.operations import Operation
+from microcosm_pubsub.conventions import created
+from werkzeug.exceptions import UnprocessableEntity
+
+from microcosm_eventsource.errors import (
+    IllegalInitialStateError,
+    IllegalStateTransitionError,
+)
+
+
+class EventController(CRUDStoreAdapter):
+
+    @property
+    def sns_producer(self):
+        return self.graph.sns_producer
+
+    def create(self, event_type, **kwargs):
+        """
+        Create an event, validating the underlying state machine.
+
+        """
+        self.validate_required_fields(event_type, **kwargs)
+        parent = self.store.retrieve_most_recent(**kwargs)
+        version, state = self.process_state_transition(parent, event_type)
+        event = self.create_event(parent, event_type, version, state, **kwargs)
+        self.publish_event(event)
+        return event
+
+    def validate_required_fields(self, event_type, **kwargs):
+        """
+        Validate type-specific required fields.
+
+        """
+        missing_required_fields = [
+            required_field
+            for required_field in event_type.value.requires
+            if kwargs.get(required_field) is None
+        ]
+        if missing_required_fields:
+            raise with_context(
+                UnprocessableEntity("Validation error"), [
+                    {
+                        "message": "Missing required field: '{}'".format(
+                            camelize(required_field, uppercase_first_letter=False),
+                        ),
+                        "field": camelize(required_field, uppercase_first_letter=False),
+                        "reasons": [
+                            "Event type '{}' requires '{}'".format(
+                                event_type.name,
+                                camelize(required_field, uppercase_first_letter=False),
+                            )
+                        ],
+                    }
+                    for required_field in missing_required_fields
+                ],
+            )
+
+    def process_state_transition(self, parent, event_type):
```

Comment:
Process a state transition by:
 - Checking the DNF syntax to determine if the transition is legal
 - Determining the state to transition to
 - Determine the version to transition to